Plugin.pm

```diff
@@ -1749,30 +1814,114 @@ sub albumInfoMenu {
 	my @artists;
 	push @artists, $album->artistsForRoles('ARTIST'), $album->artistsForRoles('ALBUMARTIST');
 
-	return _objInfoHandler( $client, $artists[0]->name, $albumTitle );
+	my $label;
+	my $labelId;
+	my $composers;
+	my $works;
+	my $qobuzAlbum;
+	my $items = [];
+
+	if ( !%$remoteMeta && $url =~ /^qobuz:/ ) {
+		my $albumId = (split /:/, $url)[-1];
+
+		getAPIHandler($client)->getAlbum(sub {
+			$qobuzAlbum = shift;
+
+			if (!$qobuzAlbum) {
+				$log->error("Get album ($albumId) failed");
+				return;
+			}
+		}, $albumId);
+
+		if ( $qobuzAlbum ) {
```

Comment:
Wouldn't that almost always fail, unless the album happened to be in the cache? This conditional block should be in the callback really, shouldn't it? If we wanted to check for cached values only, we should do so directly, not rely on an async call to return in a timely fashion by chance.

Reply:
The thing is, Slim::Menu::Albuminfo/Trackinfo triggers this (or trackinfoMenu) when building the menu containing the "On Qobuz" entry, and then again when clicking on "On Qobuz". This was existing behaviour, I didn't change this.

That's why I moved everything that relies on Qobuz metadata inside the "On Qobuz" option: by the time of that second click, it will be cached, even if it had expired before.

Reply:
Hmm... as I thought this might have been existing code, I double checked. But I still can't find it?

Reply:
Not in the plugin - but look in Slim::Menu::TrackInfo.pm:

```
	# Get plugin metadata for remote tracks
	my $remoteMeta = {};
	if ( $track->remote && blessed($client) ) {
		my $handler = Slim::Player::ProtocolHandlers->handlerForURL($url);
		if ( $handler && $handler->can('getMetadataFor') ) {
			$remoteMeta = $handler->getMetadataFor( $client, $url );
		}
	}

```
There is no equivalent in AlbumInfo.pm, which is why I added the code in the plugin function.

Reply:
I'm not talking about the caller, but about the above constructions where you do an async call to `getAlbum()`, but expect the `$qobuzAlbum` variable to be set in the callback immediately, continuing in a sync manner. If you expect the album to be in the cache already, why don't you read it from the cache directly? Relying on a callback to be called before the code continues is a bit of a lottery: it can work, but you never know.

Reply:
Ok, I see that we only cache the raw request's response. So we can't read from the cache directly. But I still believe this is wrong. It might work most of the time, because (as you said) the data will be cached. But it's asking for trouble if ever we change that code. What I'd suggest is to modify `getAlbum()` to cache the results of `_precacheAlbum`, and return that in a sync manner if available, while still running the async lookup. That way you can be sure that once the call returns you have the data if it was available. In the current situation you'd only have it if it was cached _AND_ if the async calls were fast enough to call the callback in time.

I hope this clarifies things... seems all a bit confusing, sorry.

Reply:
_If_ the cached data does exist, then you're fine. Otherwise the async HTTP request will be called, and your code continues with the check for `$qobuzAlbum` _before_ the HTTP response has arrived. LMS would continue to run, while the async call is waiting for the response. And only once that response is in, `$qobuzAlbum` would be populated - too late for the evaluation of its content.

This all might work now under normal circumstances. But if ever somebody decided to modify one of the methods you listed, the call might not return data from the cache, or might do some other async call, and you wouldn't get back `$qobuzAlbum`. Therefore I suggest you have `getAlbum()` return cached data in a sync fashion, and don't wait for the async response in that particular caller.

Reply:
So I am understanding it reasonably well, I think. What you describe is why I moved the options requiring Qobuz metadata inside the "On Qobuz" item, as described in my original PR description. This way, we get two shots, and don't need the data until the second one. Of course, the user could still be ahead of the async request, but I don't see how that can be avoided while maintaining the requirement for all calls to the Qobuz backend to be async. 

If we don't have the data cached (or the cached data has expired) we have no choice but to go async, do we?

I did test this extensively after setting the cache expiry to 5 minutes and I didn't trip it up. But I'm happy to change `getAlbum()` - just to confirm, you'd like me to copy the cache retrieval attempt from `_get()`, return if successful, otherwise carry on into `_get()`?

Reply:
What I'm saying is that this is prone to failure should the code chain leading to reading the cache change at some point. You can't rely on the callback populating `$qobuzAlbum` being called before the code continues. Understood - it happens to work now the way you expect it to. But you must not rely on a callback being called in a sync way. Just don't.

If you poke around the code you'll see that the scanner would cache the (almost) identical album information in `album_with_tracks_{albumId}`:

https://github.com/pierrepoulpe/SqueezeboxQobuz/blob/b655132c41f9e886604a646449b5c89886e6e3e7/Importer.pm#L123

Now you could use this cache key to check for the existence of cached data in a sync manner. If it didn't exist you still can do the async call to get it cached for the next call. 

```diff
diff --git a/Plugin.pm b/Plugin.pm
index 9dcb8a2..6247ab6 100644
--- a/Plugin.pm
+++ b/Plugin.pm
@@ -77,6 +77,8 @@ my $log = Slim::Utils::Log->addLogCategory( {
 use constant PLUGIN_TAG => 'qobuz';
 use constant CAN_IMAGEPROXY => (Slim::Utils::Versions->compareVersions($::VERSION, '7.8.0') >= 0);
 
+my $cache = Plugins::Qobuz::API::Common->getCache();
+
 sub initPlugin {
 	my $class = shift;
 
@@ -1818,12 +1820,13 @@ sub albumInfoMenu {
 	my $labelId;
 	my $composers;
 	my $works;
-	my $qobuzAlbum;
 	my $items = [];
 
 	if ( !%$remoteMeta && $url =~ /^qobuz:/ ) {
 		my $albumId = (split /:/, $url)[-1];
 
+		my $qobuzAlbum = $cache->get('album_with_tracks_' . $albumId);
+
 		getAPIHandler($client)->getAlbum(sub {
 			$qobuzAlbum = shift;
 
@@ -1831,7 +1834,10 @@ sub albumInfoMenu {
 				$log->error("Get album ($albumId) failed");
 				return;
 			}
-		}, $albumId);
+			else {
+				$cache->set('album_with_tracks_' . $albumId, $qobuzAlbum, time() + 86400 * 90);
+			}
+		}, $albumId) unless $qobuzAlbum;
 
 		if ( $qobuzAlbum ) {
 			my %seen;
```

I'd then move the `->getAlbum() unless $qobuzAlbum;` in the else of the following `if($qobuzAlbum)`, but didn't do this here to keep the diff short. 

Does that make sense?

Reply:
Yes, and it has helped me understand the rationale behind the code in getMetaDataFor():

```
	if ($id && $client && $client->master->pluginData('fetchingMeta')) {
		Slim::Control::Request::notifyFromArray( $client, [ 'newmetadata' ] );
		$meta = Plugins::Qobuz::API->getCachedFileInfo($id);
	}
	elsif ($id) {
		$client->master->pluginData( fetchingMeta => 1 ) if $client;

		$meta = Plugins::Qobuz::Plugin::getAPIHandler($client)->getTrackInfo(sub {
			$client->master->pluginData( fetchingMeta => 0 ) if $client;
		}, $id);
	}

```
 Should we also add a check for metadata retrieval in progress like this does?

Reply:
Actually, not quite: that code will call `getTrackInfo()` every time unless metadata retrieval by itself is actually in progress (that is, its been called multiple in quick succession), so it mostly relies on the cache retrieval at the top of `getTrackInfo()`

But similar.

Reply:
I believe the use cases aren't too similar: metadata lookup is something which can happen frequently, and even in bursts (eg. track change with many web UIs and other control clients connected). Menu navigation OTOH is mostly interactive, thus slow.

Reply:
OK, pushing changes now. 

I amended your solution slightly to ensure we don't cache unreleased albums, because the metadata is not stable before release date (as we found out before https://github.com/pierrepoulpe/SqueezeboxQobuz/pull/38/files )

I didn't move `->getAlbum()` as you suggested, for the same reason. If we did this, we'd never even attempt to populate `$qobuzAlbum` for unreleased albums, which would be a shame because the function does work most of the time even when the album is uncached.

Reply:
Thanks!